tools/hardware/memory.py

```diff
@@ -82,9 +88,13 @@ def align_base(self, align_bits):
         ''' align this region up to a given number of bits '''
         new_base = utils.align_up(self.base, align_bits)
         diff = new_base - self.base
-        new_size = self.size - diff
-        new = Region(new_base, new_size, self.owner)
-        return new
+        if self.size < diff:
+            raise ValueError(
+                'can''t align region base to {} bits, region {} too small'.format(
+                    align_bits, self))
```

Comment:
Would suggest using `f".."` strings for these:
```suggestion
                f"can't align region base to {align_bits} bits, {self} too small")
```
(Have left out `region`, because the `__str__` for `Region` will contain the word Region already)

Reply:
f-String are available in Python 3.6, so we have this version? We could merge this and then convert to f-String in a separate PR.

Reply:
Hm, I didn't actually check the python version, I just saw that we're using them in other scripts already. So I guess we must have a minimum of python 3.6. We really should set and check minimum versions for the tool chain. Happy to do this separately.